### DISCUSSION TITLE
how is graphical console supposed to work in BIOS mode?

### DISCUSSION THREAD

**Discussion by valkolaci** [ipxe/ipxe#849] [Q&A] [answered]
I have tried to utilize the graphical console ([https://ipxe.org/cmd/console](https://ipxe.org/cmd/console)).
For EFI boots, it works fine, however, when the machine boots in BIOS mode, the `console` command does nothing.
I'm playing in QEMU virtual machines for now. As the QEMU BIOS PXE code is in fact an iPXE, but with no `CONSOLE_FRAMEBUFFER` support, I first have to boot my custom-built iPXE from the network (`ipxe.lkrn`). There `console` command exists but does nothing else beyond clearing the screen (and staying at the original 80x25 resolution). I tried adding `CONSOLE_FRAMEBUFFER` + `CONSOLE_VESAFB`, and tried even undefining `CONSOLE_PCBIOS`, but nothing ever changed.

How is this supposed to work?

**Comment by NiKiZe** [comment, marked as answer]
You also need a Vesa compatible display.
You want CONSOLE_CMD and CONSOLE_FRAMEBUFFER and you should then be able to chang resolution. VESAFB is probably pulled in as needed.

Are you starting qemu manually? Could you maybe grab the command it is running with?

**Reply by valkolaci** [reply]
> You also need a Vesa compatible display. You want CONSOLE_CMD and CONSOLE_FRAMEBUFFER and you should then be able to chang resolution. VESAFB is probably pulled in as needed.

I have both `CONSOLE_CMD` and `CONSOLE_FRAMEBUFFER` defined. And obviously I do have a `console` command. Which returns an error if I make mistakes on the command line, try to add a picture with an invalid URL, and so on. And it does not return any errors when every parameter looks ok, but it also does not do the resolution change. So nothing suggests that I have any problems - it's only the result that is missing.

If I boot Linux with vga=ask, I get tons of VESA options to choose from (and they actually do work), so I'd guess that I do have a VESA-compatible display:

![vgaask](https://user-images.githubusercontent.com/26465176/211430177-2e459ea4-5622-4942-8cbf-556f79810f40.png)

> 
> Are you starting qemu manually? Could you maybe grab the command it is running with?

It's running from virt-manager, but of course I can print the running process parameters:

```
/usr/bin/qemu-system-x86_64
-name guest=linux64,debug-threads=on
-S
-object {"qom-type":"secret","id":"masterKey0","format":"raw","file":"/var/lib/libvirt/qemu/domain-68-linux64/master-key.aes"}
-machine pc-i440fx-7.1,usb=off,dump-guest-core=off,memory-backend=pc.ram
-accel kvm
-cpu qemu64
-m 2048
-object {"qom-type":"memory-backend-ram","id":"pc.ram","size":2147483648}
-overcommit mem-lock=off
-smp 2,sockets=2,cores=1,threads=1
-uuid dd819705-fffc-4010-82f2-83318cc75bb9
-no-user-config
-nodefaults
-chardev socket,id=charmonitor,fd=24,server=on,wait=off
-mon chardev=charmonitor,id=monitor,mode=control
-rtc base=utc,driftfix=slew
-global kvm-pit.lost_tick_policy=delay
-no-hpet
-no-shutdown
-global PIIX4_PM.disable_s3=1
-global PIIX4_PM.disable_s4=1
-boot menu=off,strict=on
-device {"driver":"ich9-usb-ehci1","id":"usb","bus":"pci.0","addr":"0x4.0x7"}
-device {"driver":"ich9-usb-uhci1","masterbus":"usb.0","firstport":0,"bus":"pci.0","multifunction":true,"addr":"0x4"}
-device {"driver":"ich9-usb-uhci2","masterbus":"usb.0","firstport":2,"bus":"pci.0","addr":"0x4.0x1"}
-device {"driver":"ich9-usb-uhci3","masterbus":"usb.0","firstport":4,"bus":"pci.0","addr":"0x4.0x2"}
-blockdev {"driver":"file","filename":"/var/lib/libvirt/ssdimages/linux64.qcow2","node-name":"libvirt-1-storage","auto-read-only":true,"discard":"unmap"}
-blockdev {"node-name":"libvirt-1-format","read-only":false,"driver":"qcow2","file":"libvirt-1-storage","backing":null}
-device {"driver":"ide-hd","bus":"ide.0","unit":0,"drive":"libvirt-1-format","id":"ide0-0-0"}
-netdev tap,fd=25,id=hostnet0
-device {"driver":"e1000","netdev":"hostnet0","id":"net0","mac":"08:00:27:00:00:02","bootindex":1,"bus":"pci.0","addr":"0x3"}
-chardev pty,id=charserial0
-device {"driver":"isa-serial","chardev":"charserial0","id":"serial0","index":0}
-device {"driver":"usb-tablet","id":"input0","bus":"usb.0","port":"1"}
-audiodev {"id":"audio1","driver":"none"}
-vnc 127.0.0.1:0,audiodev=audio1
-device {"driver":"VGA","id":"video0","vgamem_mb":16,"bus":"pci.0","addr":"0x2"}
-device {"driver":"virtio-balloon-pci","id":"balloon0","bus":"pci.0","addr":"0x5"}
-sandbox on,obsolete=deny,elevateprivileges=deny,spawn=deny,resourcecontrol=deny
-msg timestamp=on
```

**Reply by mcb30** [reply]
Could you try building with `DEBUG=fbcon,vesafb` to see if there are any error messages reported when you attempt to switch into a graphical console mode?

Please also show the exact `console` command that you are using, just for sanity checking!

**Reply by valkolaci** [reply]
> Could you try building with `DEBUG=fbcon,vesafb` to see if there are any error messages reported when you attempt to switch into a graphical console mode?

I added the `DEBUG` flags, but there were nothing extra printed when using `command`.

> 
> Please also show the exact console command you are using, just for sanity checking!

I tried these (they all work in ipxe.efi):
```
console -x 1024 -y 768
console -x 1024 -y 768 -d 8
console -x 1024 -y 768 -d 16
console -x 1024 -y 768 -d 24
console -x 1024 -y 768 -d 32
console -x 1280 -y 800
```
Nothing changed, no error messages, just clearing the display and returning a fresh prompt, resolution staying the same as the default (I guess it's 80x25, 640x400).

This is the build script I use:
```
#!/bin/bash
SCRIPTNAME=$(readlink -f "$0")
DIRNAME=$(dirname "$SCRIPTNAME")
SUBDIR=ipxe
BUILDDIR="$DIRNAME/$SUBDIR"
cd "$DIRNAME" || exit 1
if [ ! -d "$BUILDDIR" ]; then
  git clone https://github.com/ipxe/ipxe.git "$SUBDIR" || exit 1
  if [ ! -d "$BUILDDIR" ]; then
    echo "Cannot find build dir $BUILDDIR after git clone"
    exit 1
  fi
  cd "$BUILDDIR/src" || exit 1
else
  cd "$BUILDDIR/src" || exit 1
  git pull || exit 1
fi
mkdir -p config/local || exit 1
make clean
cp "$DIRNAME/config.h" config/local/general.h || exit 1
make NO_WERROR=1 DEBUG=fbcon,vesafb bin/undionly.kpxe bin/ipxe.lkrn || exit 1
cp bin/undionly.kpxe "$DIRNAME/" || exit 1
cp bin/ipxe.lkrn "$DIRNAME/" || exit 1
make NO_WERROR=1 bin-x86_64-efi/ipxe.efi || exit 1
cp bin-x86_64-efi/ipxe.efi "$DIRNAME/" || exit 1
```
and the config.h:
```
#undef BANNER_TIMEOUT
#define BANNER_TIMEOUT 20
#define SANBOOT_PROTO_ISCSI
#define SANBOOT_PROTO_AOE
#define SANBOOT_PROTO_IB_SRP
#define IMAGE_SCRIPT
#define NSLOOKUP_CMD
#define TIME_CMD
#define VLAN_CMD
#define REBOOT_CMD
#define POWEROFF_CMD
#define PCI_CMD
#define PARAM_CMD
#define CONSOLE_CMD
#ifdef CONSOLE_PCBIOS
#define IMAGE_ELF
#define IMAGE_MULTIBOOT
#define IMAGE_PXE
#define IMAGE_BZIMAGE
#endif
#ifdef CONSOLE_EFI
#define IMAGE_EFI
#endif
#define CONSOLE_FRAMEBUFFER
```

**Reply by mcb30** [reply]
Thank you for sharing the details of your build setup.

The problem is that you are trying to configure `CONSOLE_FRAMEBUFFER` via the wrong header file. As per https://ipxe.org/buildcfg/console_framebuffer you need to configure this via `console.h`, not `general.h`.

**Reply by valkolaci** [reply]
> Thank you for sharing the details of your build setup.
> 
> The problem is that you are trying to configure `CONSOLE_FRAMEBUFFER` via the wrong header file. As per https://ipxe.org/buildcfg/console_framebuffer you need to configure this via `console.h`, not `general.h`.

Hah! Bingo!
That was it.
Thank you.
